File: course_1/module_4/examples/randomized_contraction.py
import random
from concurrent.futures import ThreadPoolExecutor


class Graph:
    """
    Represents an undirected graph with parallel edges.
    """

    def __init__(self, input_adjacency_list):
        """
        Initializes the graph with an adjacency list.
        :param input_adjacency_list: A dictionary where keys are vertices and values are lists of adjacent vertices.
        """
        self.graph = {k: list(v) for k, v in input_adjacency_list.items()}  # Optimized shallow copy.

    def contract_edge(self, u, v):
        """
        Contracts an edge (u, v), merging vertex v into vertex u.
        :param u: One endpoint of the edge.
        :param v: The other endpoint of the edge.
        """
        # Add all edges of v to u.
        self.graph[u].extend(self.graph[v])

        # Replace all occurrences of v with u in the graph.
        for neighbor in self.graph[v]:
            self.graph[neighbor] = [u if x == v else x for x in self.graph[neighbor]]

        # Remove self-loops.
        self.graph[u] = [x for x in self.graph[u] if x != u]

        # Delete the merged vertex.
        del self.graph[v]

    def get_random_edge(self):
        """
        Selects a random edge (u, v) from the graph.
        :return: A tuple (u, v) representing an edge.
        """
        u = random.choice(list(self.graph.keys()))
        v = random.choice(self.graph[u])
        return u, v

    def get_min_cut(self):
        """
        Runs the randomized contraction algorithm to find a minimum cut.
        :return: The number of crossing edges in the minimum cut.
        """
        while len(self.graph) > 2:
            u, v = self.get_random_edge()
            self.contract_edge(u, v)

        # After contraction, count the edges between the two remaining vertices.
        remaining_vertices = list(self.graph.keys())
        return len(self.graph[remaining_vertices[0]])


class RandomizedContractionAlgorithm:
    """
    Implements the randomized contraction algorithm for finding the minimum cut in a graph.
    """

    def __init__(self, input_adjacency_list):
        """
        Initializes the algorithm with a graph represented as an adjacency list.
        :param input_adjacency_list: A dictionary where keys are vertices and values are lists of adjacent vertices.
        """
        self.original_graph = input_adjacency_list

    def find_min_cut(self, trials=None, parallel=False):
        """
        Finds the minimum cut by running the randomized contraction algorithm multiple times.
        :param trials: Number of trials to run (optional). Defaults to N^2 if not provided.
        :param parallel: Whether to run the trials in parallel.
        :return: The minimum cut found.
        """
        n = len(self.original_graph)
        if trials is None:
            trials = n * n  # Number of trials.

        min_cut = float('inf')

        def run_trial(_):
            """
            Runs a single trial of the randomized contraction algorithm.
            :param _: Ignored argument for compatibility with map function.
            :return: Minimum cut found in this trial.
            """
            graph = Graph(self.original_graph)
            return graph.get_min_cut()

        if parallel:
            # Run trials in parallel.
            with ThreadPoolExecutor() as executor:
                results = executor.map(run_trial, range(trials))
            min_cut = min(results)
        else:
            # Run trials sequentially.
            for _ in range(trials):
                min_cut = min(min_cut, run_trial(None))

        return min_cut


# Example usage.
if __name__ == "__main__":
    # Example adjacency list.
    example_adjacency_list = {
        1: [2, 3, 4],
        2: [1, 3, 4],
        3: [1, 2, 4],
        4: [1, 2, 3]
    }

    rca = RandomizedContractionAlgorithm(example_adjacency_list)
    result = rca.find_min_cut(parallel=True)  # Set parallel=True for parallel execution.
    print(f"Minimum cut: {result}")

File: course_1/module_4/examples/randomized_contraction.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// Graph represents an undirected graph with parallel edges.
type Graph struct {
	AdjacencyList map[int][]int
}

// NewGraph creates a new graph based on the given adjacency list.
func NewGraph(adjacencyList map[int][]int) *Graph {
	newList := make(map[int][]int)
	for k, v := range adjacencyList {
		newList[k] = append([]int(nil), v...)
	}
	return &Graph{AdjacencyList: newList}
}

// ContractEdge contracts two nodes (u and v) by merging them into a single node.
func (g *Graph) ContractEdge(u, v int) {
	// Add edges of v to u.
	g.AdjacencyList[u] = append(g.AdjacencyList[u], g.AdjacencyList[v]...)

	// Replace all occurrences of v with u in the graph.
	for _, neighbor := range g.AdjacencyList[v] {
		for i, val := range g.AdjacencyList[neighbor] {
			if val == v {
				g.AdjacencyList[neighbor][i] = u
			}
		}
	}

	// Remove self-loops.
	filtered := g.AdjacencyList[u][:0]
	for _, val := range g.AdjacencyList[u] {
		if val != u {
			filtered = append(filtered, val)
		}
	}
	g.AdjacencyList[u] = filtered

	// Delete vertex v from the graph.
	delete(g.AdjacencyList, v)
}

// GetRandomEdge selects a random edge (u, v) from the graph.
func (g *Graph) GetRandomEdge() (int, int) {
	vertices := make([]int, 0, len(g.AdjacencyList))
	for vertex := range g.AdjacencyList {
		vertices = append(vertices, vertex)
	}

	u := vertices[rand.Intn(len(vertices))]
	v := g.AdjacencyList[u][rand.Intn(len(g.AdjacencyList[u]))]
	return u, v
}

// GetMinCut runs the contraction algorithm to compute the minimum cut of the graph.
func (g *Graph) GetMinCut() int {
	for len(g.AdjacencyList) > 2 {
		u, v := g.GetRandomEdge()
		g.ContractEdge(u, v)
	}

	// After contraction, the remaining edges define the minimum cut.
	for _, edges := range g.AdjacencyList {
		return len(edges)
	}
	return 0
}

// RandomizedContractionAlgorithm represents the randomized contraction algorithm.
type RandomizedContractionAlgorithm struct {
	OriginalAdjacencyList map[int][]int
}

// NewRandomizedContractionAlgorithm creates a new instance of the algorithm.
func NewRandomizedContractionAlgorithm(adjacencyList map[int][]int) *RandomizedContractionAlgorithm {
	return &RandomizedContractionAlgorithm{OriginalAdjacencyList: adjacencyList}
}

// FindMinCut runs multiple trials of the contraction algorithm to find the minimum cut.
func (rca *RandomizedContractionAlgorithm) FindMinCut(trials int, parallel bool) int {
	if trials == 0 {
		n := len(rca.OriginalAdjacencyList)
		trials = n * n
	}

	minCut := int(^uint(0) >> 1) // Initialize with the maximum possible integer value.

	if parallel {
		var wg sync.WaitGroup
		resultChan := make(chan int, trials)

		// Run trials in parallel.
		for i := 0; i < trials; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				graph := NewGraph(rca.OriginalAdjacencyList)
				resultChan <- graph.GetMinCut()
			}()
		}

		// Wait for all trials to complete and collect results.
		wg.Wait()
		close(resultChan)

		for result := range resultChan {
			if result < minCut {
				minCut = result
			}
		}
	} else {
		// Run trials sequentially.
		for i := 0; i < trials; i++ {
			graph := NewGraph(rca.OriginalAdjacencyList)
			if cut := graph.GetMinCut(); cut < minCut {
				minCut = cut
			}
		}
	}

	return minCut
}

func main() {
	// Example adjacency list.
	adjacencyList := map[int][]int{
		1: {2, 3, 4},
		2: {1, 3, 4},
		3: {1, 2, 4},
		4: {1, 2, 3},
	}

	rand.Seed(time.Now().UnixNano())

	// Create an instance of the algorithm and find the minimum cut.
	rca := NewRandomizedContractionAlgorithm(adjacencyList)
	minCut := rca.FindMinCut(0, true) // 0 to calculate trials, true for parallel execution.
	fmt.Printf("Minimum cut: %d\n", minCut)
}
